### PR TITLE
8302462: [REDO] 8297487: G1 Remark: no need to keep alive oop constants of nmethods on stack

### PR DESCRIPTION
**Pull request by reinrich**
This is a REDO of JDK-8297487 after the issues that cause the BACKOUT were resolved with [JDK-8300915](https://bugs.openjdk.org/browse/JDK-8300915)

Testing:
* The reproducer `test/langtools:tier1` was executed without crash for 12h.
* Multiple iterations of jtreg tests tiers 1-4  on standard platforms and also on ppc64le.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302462](https://bugs.openjdk.org/browse/JDK-8302462): [REDO] 8297487: G1 Remark: no need to keep alive oop constants of nmethods on stack

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [e423ed1a](https://git.openjdk.org/jdk/pull/12561/files/e423ed1ab181bfdf0fb402b1dfbab8d20bead89d)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12561/head:pull/12561` \
`$ git checkout pull/12561`

Update a local copy of the PR: \
`$ git checkout pull/12561` \
`$ git pull https://git.openjdk.org/jdk pull/12561/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12561`

View PR using the GUI difftool: \
`$ git pr show -t 12561`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12561.diff">https://git.openjdk.org/jdk/pull/12561.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shared/barrierSet.cpp [hunk 1]
```diff
@@ -57,8 +57,11 @@ static BarrierSetNMethod* select_barrier_set_nmethod(BarrierSetNMethod* barrier_
     // The GC needs nmethod entry barriers to do concurrent GC
     return barrier_set_nmethod;
   } else {
-    // The GC needs nmethod entry barriers to deal with continuations
-    // and code cache unloading
+    // The GC needs nmethod entry barriers for code cache unloading.
+    // Concurrent GC needs them also for preserving the weakly referenced objects in
+    // the constant pool of nmethods as part of the SATB snapshot, and to deal with
+    // compiled frames contained in continuation stack chunks allocated after
+    // concurent mark start.
```
Comment (reinrich) [inline src/hotspot/share/gc/shared/barrierSet.cpp:64]: I've adapted the comment after https://mail.openjdk.org/pipermail/hotspot-gc-dev/2023-January/041194.html

**Review by tschatzl — APPROVED** [review]
lgtm

File: src/hotspot/share/gc/shared/barrierSet.cpp (hunk 1 above)
Comment (albertnetymk) [reply src/hotspot/share/gc/shared/barrierSet.cpp:64]: I feel the comment is at the wrong level of abstraction, compared with the surrounding code.

Maybe just `Provide nmethod entry barrier to support concurrent marking and code cache unloading.` at the top of the whole method.

The more elaborate "inside" about nmethod-entry-barrier probably makes more sense at `BarrierSetNMethod::nmethod_entry_barrier`.

Comment (reinrich) [reply src/hotspot/share/gc/shared/barrierSet.cpp]: You are right. I've reverted that change and added a line of comment to `BarrierSetNMethod::nmethod_entry_barrier`

**Review by albertnetymk — APPROVED** [review]

**Comment by reinrich** [comment]
Thanks for the reviews.

/integrate

**Pull request closed without merging**